kong/core/router.lua

```diff
@@ -128,18 +143,48 @@ local function marshall_api(api)
     if #api.uris > 0 then
       api_t.match_rules = bor(api_t.match_rules, MATCH_RULES.URI)
 
-      for i, uri in ipairs(api.uris) do
-        local escaped_uri = [[\Q]] .. uri .. [[\E]]
-        local strip_regex = escaped_uri .. [[\/?(.*)]]
+      for _, uri in ipairs(api.uris) do
+        if re_match(uri, [[^[a-zA-Z0-9\.\-_~/%]*$]]) then
+          -- plain URI or URI prefix
+          local escaped_uri = [[\Q]] .. uri .. [[\E]]
+          local strip_regex = escaped_uri .. [[/?(?P<stripped_uri>.*)]]
+
+          api_t.uris[uri] = {
+            prefix        = true,
+            strip_regex   = strip_regex,
+          }
+
+          insert(api_t.uris, {
+            prefix      = true,
+            value       = uri,
+            regex       = escaped_uri,
+            strip_regex = strip_regex,
+          })
+
+        else
+          -- regex URI
+          local strip_regex = uri .. [[/?(?P<stripped_uri>.*)]]
+
+          local from, _, err = re_find(uri, [[\(.*\)]], "jo")
```

Comment:
this expression needs some more fleshing out. for example, this matches the literal string

`\(foo\)`

which is _not_ a capture group. additionally, the `*` quantifier results in greedy + backtracking behavior; we can use a more performant negative char class (`[^)]+`). 

as an alternative, consider the following expression (untested, but passing for this scenario):

`(?<!\\)\([^)]+(?<!\\)\)`

for readability, we may want to consider using char groups instead of escaping `\`, as well as the `x` opt.